src/algebra/category/Module/change_of_rings.lean

```diff
@@ -161,4 +166,166 @@ variables {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S] (f : R →
+  begin
+    ext,
+    rw [hom_equiv.to_restrict_scalars_apply, hom_equiv.from_extend_scalars_apply, lift.tmul,
+      linear_map.coe_mk, linear_map.coe_mk],
+    convert one_smul _ _,
+  end }
+
+/--
+For any `R`-module X, there is a natural `R`-linear map from `X` to `X ⨂ S` by sending `x ↦ x ⊗ 1`
+-/
+@[simps] def unit.map {X} : X ⟶ (extend_scalars f ⋙ restrict_scalars f).obj X :=
+{ to_fun := λ x, (1 : S) ⊗ₜ[R, f] x,
+  map_add' := λ x x', by { rw tensor_product.tmul_add, },
+  map_smul' := λ r x, by { letI m1 : module R S := module.comp_hom S f, tidy } }
+
+/--
+The natural transformation from identity functor on `R`-module to the composition of extension and
+restriction of scalars.
+-/
+@[simps] def unit : 𝟭 (Module R) ⟶ extend_scalars f ⋙ restrict_scalars f :=
+{ app := λ _, unit.map f, naturality' := λ X X' g, by tidy }
+
+/--
+For any `S`-module Y, there is a natural `R`-linear map from `S ⨂ Y` to `Y` by
+`s ⊗ y ↦ s • y`
+-/
+@[simps] def counit.map {Y} : (restrict_scalars f ⋙ extend_scalars f).obj Y ⟶ Y :=
+begin
+  letI m1 : module R S := module.comp_hom S f,
+  letI m2 : module R Y := module.comp_hom Y f,
+  refine ⟨tensor_product.lift ⟨λ (s : S), ⟨λ (y : Y), s • y, smul_add _, _⟩, _, _⟩, _, _⟩,
+  { intros, rw [ring_hom.id_apply, restrict_scalars.smul_def, ←mul_smul, mul_comm, mul_smul,
+      restrict_scalars.smul_def], },
+  { intros, ext, simp only [linear_map.add_apply, linear_map.coe_mk, add_smul], },
+  { intros, ext,
+    simpa only [ring_hom.id_apply, linear_map.smul_apply, linear_map.coe_mk,
+      @restrict_scalars.smul_def _ _ _ _ f ⟨S⟩, smul_eq_mul, mul_smul], },
+  { intros, rw [map_add], },
+  { intros s z,
+    rw [ring_hom.id_apply],
+    induction z using tensor_product.induction_on with x s' z1 z2 ih1 ih2,
+    { simp only [smul_zero, map_zero], },
+    { simp only [extend_scalars.smul_tmul, linear_map.coe_mk, tensor_product.lift.tmul, mul_smul] },
+    { rw [smul_add, map_add, map_add, ih1, ih2, smul_add], } },
+end
+
+/--
+The natural transformation from the composition of restriction and extension of scalars to the
+identity functor on `S`-module.
+-/
+@[simps] def counit : (restrict_scalars f ⋙ extend_scalars f) ⟶ (𝟭 (Module S)) :=
+{ app := λ _, counit.map f,
+  naturality' := λ Y Y' g,
+    begin
+      ext z, induction z using tensor_product.induction_on,
+      { simp only [map_zero] },
+      { simp only [category_theory.functor.comp_map, Module.coe_comp, function.comp_app,
+          extend_scalars.map_tmul, restrict_scalars.map_apply, counit.map_apply, lift.tmul,
+          linear_map.coe_mk, category_theory.functor.id_map, linear_map.map_smulₛₗ,
+          ring_hom.id_apply] },
+      { simp only [map_add, *] },
+    end }
+
+end extend_restrict_scalars_adj
+
+/--
+Given commutative rings `R, S` and a ring hom `f : R →+* S`, the extension and restriction of
+scalars by `f` are adjoint to each other.
+-/
+def extend_restrict_scalars_adj {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S]
```

Comment:
```suggestion
@[simps]
def extend_restrict_scalars_adj {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S]
```

Because of how you named the auxiliary definitions above, there might be name conflicts.